src/window/editor/select_city_by_type.c

```diff
@@ -90,15 +89,22 @@ static void populate_list(int offset)
     }
 }
 
+static void close(void)
+{
+    window_go_back();
+}
+
 static void init(void (*callback)(int), empire_city_type type)
 {
     data.callback = callback;
     data.filter_type = type;
-    memset(data.list, 0, sizeof(data.list));
-    memset(data.valid_city_ids, 0, sizeof(data.valid_city_ids));
     data.list_size = 0;
+    memset(data.list, 0, sizeof(data.list));
 
     int city_array_size = empire_city_get_array_size();
+    data.valid_city_ids = malloc((city_array_size + 1) * sizeof(int));
```

Comment:
Careful here: `malloc`ing without `free`ing will cause memory leaks.

What you need is a new static var `data.valid_city_ids_size` and if `city_array_size` is larger than it, you `realloc` `data.valid_city_ids` and set `data.valid_city_ids_size` to `city_array_size`.

Reply:
Do you always need to free if you did malloc?
Or can I in this case leave the data there in memory? (Or is that bad coding practice even if it is a small amount?)

Reply:
Yes you can leave it in memory, it's a small amount to worry and it will be automatically deallocated when the game exits.